### PR TITLE
type manifest can be dir of manifests

### PR DESCRIPTION
**Pull request by 4nd3r**
All my stuff is organized around types and some types have very long manifest (~500 lines). Sure, split it up, you say, but same time having this small simple way to split up manifest into smaller manifests makes life little easier. Everything seems to work as is, with no breakage (let's see what tests do, hehe), but what do you think, @sideeffect42?

In logs:
```
VERBOSE: media.internal: Preparing object __host_media/
VERBOSE: media.internal: Running manifest and explorers for __host_media/
VERBOSE: media.internal: Running type explorers for <CdistType __host_media
VERBOSE: media.internal: Running type manifest /tmp/tmp2ms_u4n5/0480c6f84cb1b9543f0cd34655244113/data/conf/type/__host_media/manifest/base for object __host_media/
VERBOSE: media.internal: Running type manifest /tmp/tmp2ms_u4n5/0480c6f84cb1b9543f0cd34655244113/data/conf/type/__host_media/manifest/funny for object __host_media/
VERBOSE: media.internal: Running type manifest /tmp/tmp2ms_u4n5/0480c6f84cb1b9543f0cd34655244113/data/conf/type/__host_media/manifest/isofiles for object __host_media/
VERBOSE: media.internal: Running type manifest /tmp/tmp2ms_u4n5/0480c6f84cb1b9543f0cd34655244113/data/conf/type/__host_media/manifest/nginx for object __host_media/
VERBOSE: media.internal: Running type manifest /tmp/tmp2ms_u4n5/0480c6f84cb1b9543f0cd34655244113/data/conf/type/__host_media/manifest/samba for object __host_media/
VERBOSE: media.internal: Running type manifest /tmp/tmp2ms_u4n5/0480c6f84cb1b9543f0cd34655244113/data/conf/type/__host_media/manifest/srv for object __host_media/
VERBOSE: media.internal: Running type manifest /tmp/tmp2ms_u4n5/0480c6f84cb1b9543f0cd34655244113/data/conf/type/__host_media/manifest/wh40k for object __host_media/
VERBOSE: media.internal: Running type manifest /tmp/tmp2ms_u4n5/0480c6f84cb1b9543f0cd34655244113/data/conf/type/__host_media/manifest/yarr for object __host_media/
```

### CODE REVIEW

**Comment by sideeffect42** [comment]
Let me start with the positive first: I totally agree that we could do better with splitting up long scripts into separate modules.
Also in the upstream sets we have manifests (but also gencode scripts) which are essentially one giant `case ${os}` and which maybe would become more readable if they were split up, e.g. one file for each OS we support.

But then I have to say that this solution looks quite "Ansible" to me.
The creation of files in a "magic" location changes the execution (suddenly the `manifest` script is not executed anymore?)

Yes, it's not that bad, I'm nitpicking here, but I wonder if just having a stub `manifest` like the following would be worse?
```sh
#!/bin/sh -e

for manifest in "${__type:?}"/manifest/*  # or files/manifest/*
do
    "${manifest}"  # or alternatively: source manifest
done
```

It's explicit, requires no changes in the runtime, could also be applied to gencode scripts, and is more flexible:
- the different parts could be run in every order the author wishes (by listing them explicitly instead of `for`)
- some parts could be run only under a certain condition (e.g. based on OS)
- theoretically one could structure manifests even further (e.g. `manifest/generic/*`, followed by `manifest/by-os/*`, and `manifest/finally`, or whatever the author desires).

**Comment by 4nd3r** [comment]
Let's take my idea even further. We have `.skonfig/manifest/init` as default manifest. In that sense it makes more sense to have `$__type/manifest/init` as default type manifest location. And now we have even more ways to improve user experience:

1. Default manifest is `$__type/manifest`.
2. If that path is dir, then look for `$__type/manifest/init`.
3. If `$__type/manifest/init` is not there, but dir is, then iterate over `$__type/manifest/*`.

This way we:
1. Maintain compatibility with existing, duh.
2. Provide clear separation of manifests for those, who wish to split them up.
5. Provide explicit control with `$__type/manifest/init` and magic with `$__type/manifest/*`.

I don't like Ansible, but I love when tools do common simple stuff for me, without me having to think about it.

Sounds too complicated? IMHO it's not, it feels pretty comfy. We have such separation for explorers within type, and parameters and files, all automagical, but manifest does not have such logic.

**Comment by 4nd3r** [comment]
@sideeffect42 ping?

**Review by sideeffect42 — APPROVED** [review]
Let's give it a try.

Apologies for letting this sit for so long. To be honest I was very sceptical about adding new ways to do something which was already possible with very little additional code.
I'm a big believer in the "UNIX philosophy"; I see skonfig mainly as an "enabler". So if something is already possible I don't really see a reason to add new ways to achieve the same goal. Just like we removed support for processing multiple hosts in a single command line when we migrated from `cdist(1)` to `skonfig(1)`, because one could use e.g. `xargs(1)` to handle multiple hosts.

After looking back to the days when I used Puppet I remembered into how many classes a Puppet module is split up.
This approach does have its pros and cons. One one hand it makes the single classes smaller and thus easier to read and things like parameters or actions specific to an OS can be separated into their own file, but on the other hand it can make it harder to follow what a manifest does without opening all the files.

Anyway, long story short, let's merge this, give it a try and see where it takes us.

My only wish would be to also allow separation of `gencode-*` scripts, just to be consistent.

**Comment by 4nd3r** [comment]
> Anyway, long story short, let's merge this, give it a try and see where it takes us.

Been using that for month now, works fine :)

I'll take a look at gencodes next.
